mypy/constraints.py

```diff
@@ -1138,8 +1138,13 @@ def find_matching_overload_item(overloaded: Overloaded, template: CallableType)
             item, template, is_compat=mypy.subtypes.is_subtype, ignore_return=True
         ):
             return item
-    # Fall back to the first item if we can't find a match. This is totally arbitrary --
-    # maybe we should just bail out at this point.
+    # Try to return the first item with the correct self type (fixes issue 14943).
+    for item in items:
+        if isinstance(item.definition, FuncDef) and isinstance(item.definition.type, CallableType):
+            if item.bound_args and item.definition.type.arg_types:
```

Comment:
The `definition` attribute is only supposed to be used for error messages, and it's not guaranteed to be set. It cannot be reliably used for type checking. It's not serialized, in particular, but there may well be other contexts where it's lost that I'm not aware of.

Also, the `bound_args` attribute is unfortunately not documented (or the documentation seems incorrect). Again, I'm not sure if we can expect it to be always set.

Here's an alternative approach that might work. When we bind the self type to an overloaded method (in `mypy.typeops.bind_self`), what about filtering the overload items at that point to those matching the self type (only when explicit self types are used)? This filtering would happen in an earlier stage, and I believe we have access to all the necessary information without using undocumented or inconsistently set attributes. The result of the filtering could be a non-overloaded callable type or an overloaded type with fewer items.

Reply:
Thank you for your feedback.  Using two fragile attributes in five new lines - what a negative hit rate...

Your suggestion targets code sections I did not encounter yet.  My first naive implementation of your suggestion solved the original issue but broke many other things.  So far, I could not find any information to distinguish implicit and explicit self types within `bind_self`.  Is there some obvious way I am missing?

I would take a second, more thorough try within the next few days.  Please do not hesitate to fix this yourself if you already have a good solution in mind and if you want that issue fixed before the 1.2 release.

Reply:
I tried to filter within `bind_self` as suggested.  All tests pass (on my computer).  Let's see what Mypy primer says.

I am still not overly happy with my approach for the following reasons:

1. I added the `ignore_type_vars` option to `is_subtype`, which seems like a huge change for such a specific problem to me.
2. I found no way to distinguish explicit self types from implicit self types.
3. The result is still always an overloaded type (changing this within `bind_self` would be simple but requires changing its signature and eventually many code sections that use `bind_self`).

Reply:
> 3\. The result is still always an overloaded type (changing this within `bind_self` would be simple but requires changing its signature and eventually many code sections that use `bind_self`).

I played around with returning a single non-overloaded callable for a few minutes.  Maybe it's not a good idea because it changes the signatures in error reports and notes.  On the other hand, I did not encounter a case where an overloaded type with a single item caused a problem (but I did not actively search for such a problem).